targets/target_persistent.c

```diff
@@ -14,6 +21,39 @@ static int enumerate_ranges(const GumRangeDetails* details, gpointer user_data)
     return 0;
 }
 
+static void prefetch_log(const GumEvent* event, GumCpuContext* cpu_context, gpointer user_data) {
+    switch (event->type) {
+    case GUM_COMPILE: {
+        const GumCompileEvent* compile = &event->compile;
+        if (prefetch_compiled != NULL) {
+            g_hash_table_add(prefetch_compiled, compile->begin);
+        }
+    } break;
+    }
+}
+
+static void prefetch_write(int fd, GHashTable *table) {
+    GHashTableIter iter;
+    gpointer key, value;
+    g_hash_table_iter_init(&iter, table);
+    while (g_hash_table_iter_next(&iter, &key, &value)) {
+        int status = write(fd, &key, sizeof(key));
+        g_assert_cmpint(status, ==, 0);
+    }
+}
+
+static void prefetch_read(int fd, GHashTable *table) {
+    gpointer block_address;
+    while (read(fd, &block_address, sizeof(block_address)) == sizeof(block_address)) {
```

Comment:
If you prefer the pipes, you could consider reading a buffer full of address in one go to reduce the number of syscalls. This could be further reduced if prefetching is only triggered every 'n' forks. Whilst slower during the initial run-up, this will be quicker later on when there are few new paths being discovered.